go-controller/pkg/ovn/controller/services/load_balancer.go

```diff
@@ -0,0 +1,456 @@
+	for _, node := range nodeInfos {
+		nodeSwitches = append(nodeSwitches, node.switchName)
+		if globalconfig.Gateway.Mode == globalconfig.GatewayModeShared && node.gatewayRouterName != "" {
+			nodeRouters = append(nodeRouters, node.gatewayRouterName)
+		}
+	}
+
+	cbp := configsByProto(configs)
+
+	out := []ovnlb.LB{}
+	for _, proto := range protos {
+		cfgs, ok := cbp[proto]
+		if !ok {
+			continue
+		}
+		lb := ovnlb.LB{
+			Name:        makeLBName(service, proto, "cluster"),
+			Protocol:    string(proto),
+			ExternalIDs: util.ExternalIDsForObject(service),
+			Opts:        lbOpts(service),
+
+			Switches: nodeSwitches,
+			Routers:  nodeRouters,
+		}
+
+		for _, config := range cfgs {
+			if config.routerLocalMode {
+				panic("coding error: local endpoints only for cluster-wide lbConfig")
+			}
+
+			v4targets := make([]ovnlb.Addr, 0, len(config.eps.V4IPs))
+			for _, tgt := range config.eps.V4IPs {
+				v4targets = append(v4targets, ovnlb.Addr{
+					IP:   tgt,
+					Port: config.eps.Port,
+				})
+			}
+
+			v6targets := make([]ovnlb.Addr, 0, len(config.eps.V6IPs))
+			for _, tgt := range config.eps.V6IPs {
+				v6targets = append(v6targets, ovnlb.Addr{
+					IP:   tgt,
+					Port: config.eps.Port,
+				})
+			}
+
+			rules := make([]ovnlb.LBRule, 0, len(config.vips))
+			for _, vip := range config.vips {
+				if vip == "node" {
+					panic("coding error: node IP for cluster LB")
+				}
+				targets := v4targets
+				if strings.Contains(vip, ":") {
+					targets = v6targets
+				}
+
+				rules = append(rules, ovnlb.LBRule{
+					Source: ovnlb.Addr{
+						IP:   vip,
+						Port: config.inport,
+					},
+					Targets: targets,
+				})
+			}
+			lb.Rules = append(lb.Rules, rules...)
+		}
+
+		out = append(out, lb)
+	}
+	return out
+}
+
+// buildPerNodeLBs takes a list of lbConfigs and expands them to one LB per protocol per node
+// This works differently based on whether or not we're in shared or local gateway mode.
+//
+// For local gateway, per-node lbs are created for:
+// - nodePort services, attached to each node's switch, vips are node's physical IPs
+//
+// For shared gateway, per-node lbs are created for
+// - clusterip services with host-network services are attached to each node's gateway router + switch
+// - nodeport services are attached to each node's gateway router + switch, vips are node's physical IPs
+// - any services with host-network endpoints
+// - services with external IPs / LoadBalancer Status IPs
+//
+// HOWEVER, we need to replace, on each nodes gateway router only, any host-network endpoints with a special loopback address
+// see https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/design/host_to_services_OpenFlow.md
+// This is for host -> serviceip -> host hairpin
+//
+// For ExternalTrafficPolicy, all "External" IPs (NodePort, ExternalIPs, Loadbalancer Status) have:
+// - targets filtered to only local targets
+// - SkipSNAT enabled
+// This results in the creation of an additional load balancer.
```

Comment:
nit: ....additional load balancer on the node's GR.

Reply:
fixed.